File: RaccoonChat/Themes/ViewControllers/ThemesViewController.m
//
//  ThemesViewController.m
//  RaccoonChat
//
//  Created by Анна Якусевич on 03/03/2019.
//  Copyright © 2019 Hanna Yakusevych. All rights reserved.
//

#import "ThemesViewController.h"

@interface ThemesViewController()

@end

@implementation ThemesViewController

@synthesize model = _model;
@synthesize delegate = _delegate;

// MARK: - Actions
- (IBAction) didSelectTheme:(UIButton*)sender {
  NSUInteger index = [_themeButtons indexOfObject:sender];
  UIColor* backgroundColor;
  switch (index) {
    case 0:
      backgroundColor = _model.theme1;
      break;
    case 1:
      backgroundColor = _model.theme2;
      break;
    case 2:
      backgroundColor = _model.theme3;
      break;
    default:
      [NSException raise:@"wrong button idx" format:@""];
      break;
  }
  
  if (_delegate) {
    [_delegate themesViewController:self didSelectTheme:backgroundColor];
  }
}

- (IBAction)dismiss:(UIBarButtonItem *)sender {
  [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
  [super viewDidLoad];
  [_themeButtons[0] setBackgroundColor:[UIColor colorWithWhite:1 alpha:0.8]];
  [_themeButtons[1] setBackgroundColor:[UIColor colorWithWhite:1 alpha:0.8]];
  [_themeButtons[2] setBackgroundColor:[UIColor colorWithWhite:1 alpha:0.8]];
    // Do any additional setup after loading the view.
}

- (void)viewDidLayoutSubviews {
  [super viewDidLayoutSubviews];
  
}

// MARK: - Getters
- (Themes*) model {
  return _model;
}

- (id<ThemesViewControllerDelegate>)delegate {
  return _delegate;
}

// MARK: - Setters
- (void)setModel:(Themes *)newModel {
  if (_model != newModel) {
    id oldModel = _model;
    _model = [newModel retain];
    [oldModel release];
  }
}

- (void)setDelegate:(id<ThemesViewControllerDelegate>)delegate {
  if (_delegate != delegate) {
    id oldDelegate = _delegate;
    _delegate = [delegate retain];
    [oldDelegate release];
  }
}



/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (void)dealloc {
  [_model release];
  [_themeButtons release];
  [super dealloc];
}
@end

File: RaccoonChat/Themes/ViewControllers/ThemesViewController.swift
//
//  ThemesViewController.swift
//  RaccoonChat
//
//  Created by Анна Якусевич on 04/03/2019.
//  Copyright © 2019 Hanna Yakusevych. All rights reserved.
//

import UIKit

class ThemesViewController: UIViewController {

  var model: Themes!
  var changeColor: ((UIColor) -> Void)?

  @IBOutlet var themeButtons: [UIButton]!

  @IBAction func didSelectTheme(_ sender: UIButton) {
    guard
      let model = model,
      let index = themeButtons.index(of: sender) else { return }

    let backgroundColor: UIColor

    switch index {
    case 0:
      backgroundColor = model.theme1
    case 1:
      backgroundColor = model.theme2
    case 2:
      backgroundColor = model.theme3
    default:
      fatalError("The button index is out of range")
    }
    if let changeColor = changeColor {
      changeColor(backgroundColor)
    } else {
      Logger.write("Error: The closure for changing color wasn't passed")
    }

  }

  @IBAction private func dismiss(_ sender: UIBarButtonItem) {
    self.dismiss(animated: true, completion: nil)
  }

  override func viewDidLoad() {
    super.viewDidLoad()
    themeButtons[0].backgroundColor = UIColor.white.withAlphaComponent(0.8)
    themeButtons[1].backgroundColor = UIColor.white.withAlphaComponent(0.8)
    themeButtons[2].backgroundColor = UIColor.white.withAlphaComponent(0.8)
    self.view.backgroundColor = ThemeManager.currentTheme().mainColor
        // Do any additional setup after loading the view.
  }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
